# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [22]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [23]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    # apply set() function to get rid of duplicate words
    vocab_list = tuple(set(text))
    
    # Converting into dictionaries
    int_to_vocab = dict(enumerate(vocab_list))
    vocab_to_int = {ch: ii for ii, ch in int_to_vocab.items()}
    
    # return tuple
    return vocab_to_int, int_to_vocab

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


In [24]:
print('Total words in text        :{}'.format(len(tuple(text))))
print()
print('Total unique words in text :{}'.format(len(tuple(set(text)))))
print()
print('Examples of text           :\n{}'.format(text[100:2000]))

Total words in text        :3471464

Total unique words in text :72

Examples of text           :
out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! you wanna go to sleep, you wanna get up, you wanna go out again tomorrow, right? where ever you are in life, its my feeling, youve gotta go. 

jerry: (pointing

### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [25]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    tokens_dict = {'.': '<Period>',
                   ',': '<Comma>',
                   '"': '<Quotation_Mark>',
                   ';': '<Semicolon>',
                   '!': '<Exclamation_Mark>',
                   '?': '<Question_Mark>',
                   '(': '<Left_Parentheses>',
                   ')': '<Right_Parentheses>',
                   '-': '<Dash>',
                   '\n': '<Return>'}
    return tokens_dict

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [26]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [28]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [29]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
   #batch_size_total = batch_size * sequence_length
    # this calculates the complete total batches we can make
    n_batches = len(words)//batch_size
        
    # keep only enough characters to make full batches
    words = words[:n_batches * batch_size]
    
    # set as empty list
    features, target = [], []
    
    # iterate through the sequence_length at a time
    for idx in range(0, len(words) - sequence_length):
        # tensor features
        list_features = words[idx:idx + sequence_length]
        # tensor targets
        list_target = words[idx + sequence_length]   
        
        features.append(list_features)
        target.append(list_target)

    # create & batch Tensor datasets using DataLoader
    data = TensorDataset(torch.from_numpy(np.asarray(features,np.int64)), torch.from_numpy(np.asarray(target,np.int64)))
    data_loader = DataLoader(data, shuffle=True, batch_size=batch_size, drop_last=True)
    
    return data_loader

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own
test = range(2000)
testing = batch_data(test, sequence_length=8, batch_size=6)
data_iter = iter(testing)
test_x, test_y = data_iter.next()

print(test_x.shape)
print(test_x)
print()
print(test_y.shape)
print(test_y)

torch.Size([6, 8])
tensor([[  216,   217,   218,   219,   220,   221,   222,   223],
        [ 1824,  1825,  1826,  1827,  1828,  1829,  1830,  1831],
        [  165,   166,   167,   168,   169,   170,   171,   172],
        [ 1960,  1961,  1962,  1963,  1964,  1965,  1966,  1967],
        [ 1835,  1836,  1837,  1838,  1839,  1840,  1841,  1842],
        [  161,   162,   163,   164,   165,   166,   167,   168]])

torch.Size([6])
tensor([  224,  1832,   173,  1968,  1843,   169])


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [30]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 18,  19,  20,  21,  22],
        [ 24,  25,  26,  27,  28],
        [ 42,  43,  44,  45,  46],
        [ 44,  45,  46,  47,  48],
        [  9,  10,  11,  12,  13],
        [  5,   6,   7,   8,   9],
        [  4,   5,   6,   7,   8],
        [ 20,  21,  22,  23,  24],
        [  8,   9,  10,  11,  12],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 23,  29,  47,  49,  14,  10,   9,  25,  13,  12])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [31]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
         # set class variables
        self.dropout = dropout
        self.output_size = output_size
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        
        # define embedding layer        
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        
        ## define the LSTM
        # self.LSTM(input_size, n_hidden, n_layers, dropout=dropout, batch_first=True)
        # Note batch_first is default to True because when we created our batch data, the first dim is the batch size,
        # rather than the sequence length
        # Next, a dropout layer to go between LSTM and the final linear layer
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout=dropout, batch_first=True)
        
        # Define the final, fully-connected output layer
        # Takes in LSTM output which are going to be the dimension and hidden
        self.fc = nn.Linear(hidden_dim, output_size)
            
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function   
        batch_size = nn_input.size(0)

        # embeddings and lstm_out
        lstm_out, hidden = self.lstm(self.embedding(nn_input), hidden)
        
        # stack up lstm outputs using view
        lstm_out = lstm_out.contiguous().view(-1, self.hidden_dim)
        
        # dropout and fully-connected layer
        out = self.fc(lstm_out)
        
        # reshape into (batch_size, seq_length, output_size)
        out = out.view(batch_size, -1, self.output_size)
        # get last batch
        out = out[:, -1]

        return out, hidden    
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        weight = next(self.parameters()).data
        
        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                  weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        
        return hidden 
      
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [32]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # if GPU is available, we moving these inputs and targets to our GPU device
    if(train_on_gpu):
        inp, target = inp.cuda(), target.cuda()  
    
    # creating new variables for the hidden state, otherwise
    # we'd backprop through the entire training history
    # perform backpropagation and optimization
    h = tuple([each.data for each in hidden])

    # zero accumulated gradients
    rnn.zero_grad()
    # rnn.zero_grad() sets the gradients of all its parameters (including parameters of submodules) to zero.
    # If you call optim.zero_grad() that will do the same,but for all parameters that have been specified to be optimised. 
    # If you are using only rnn.parameters() in your optimiser, e.g. optim = Adam(net.parameters(), lr=1e-3), 
    # then both are equivalent, since they contain the exact same parameters.
    
    # getting the output from the model
    output, h = rnn(inp, h)
    
    # calculate loss
    loss = criterion(output, target)
    
    # perform backprop
    loss.backward()
    
    # 'clip_grad_norm' helps prevent the exploding gradient problem in RNNs / LSTMs
    # for simplicity, we set this as 5 for now
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)

    # perform a single optimization step (parameter update)
    optimizer.step()
    
    # return the loss over a batch and the hidden state produced by our model 
    return loss.item(), h


# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [33]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [43]:
# Data params
# Sequence Length
sequence_length = 10   # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [46]:
# Training parameters
# Number of Epochs
num_epochs = 15
# Learning Rate
learning_rate = 0.001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 200
# Hidden Dimension
hidden_dim = 512
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 1000

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 15 epoch(s)...
Epoch:    1/15    Loss: 5.007475462436676

Epoch:    1/15    Loss: 4.436967253684998

Epoch:    1/15    Loss: 4.2806114320755

Epoch:    1/15    Loss: 4.185076273202896

Epoch:    1/15    Loss: 4.11888398694992

Epoch:    1/15    Loss: 4.066962728738785

Epoch:    2/15    Loss: 3.925467912864879

Epoch:    2/15    Loss: 3.817287394285202

Epoch:    2/15    Loss: 3.8125609588623046

Epoch:    2/15    Loss: 3.7891331360340117

Epoch:    2/15    Loss: 3.8076311576366426

Epoch:    2/15    Loss: 3.7839770019054413

Epoch:    3/15    Loss: 3.6752729872620202

Epoch:    3/15    Loss: 3.5772882437705995

Epoch:    3/15    Loss: 3.5882576694488524

Epoch:    3/15    Loss: 3.5927019147872925

Epoch:    3/15    Loss: 3.6093140001296997

Epoch:    3/15    Loss: 3.6151821200847625

Epoch:    4/15    Loss: 3.5091524153220943

Epoch:    4/15    Loss: 3.397351525068283

Epoch:    4/15    Loss: 3.4386048262119293

Epoch:    4/15    Loss: 3.449031736373901

Epoch:    4/15   

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** 

============================================================================================================
<br>In this hyperparams tuning exercise, I keep all these metrics constant while training on one specific parameter at a time.
    - num_epochs = 10
    - batch_size = 128
    - learning_rate = 0.001
    - embedding_dim = 150
    - hidden_dim = 512
    - n_layers = 2
My first epoch started around 5.~ and progressed downward as the epoch increases. But I noticed extensive amount of time is taken in this process despite GPU-enabled. I stopped at iteration 6/7th whenever I saw the loss exhibits marginal diminishing return.
<br>============================================================================================================
- __<font color='red'><u>sequence_length (10, 20, 30)</u></font>__
        - Sequence length 10 gives us a slightly better result.
             ________________________________
             | Sequence Length    |    Loss |
             |____________________|_________|
             | 10                 |   3.541 |
             | 20                 |   3.582 |
             | 30                 |   3.899 |
             |____________________|_________|

- __batch_size (64, 128, 256)__
        - it's advisable to keep the batch size between 64 and 256. 
        - Note, higher batch size increases computational resources.
        - Batch size = 128 gives better result.
             ________________________________
             | Batch Size         |    Loss |
             |____________________|_________|
             |  64                |   3.803 |
             | 128                |   3.642 |
             | 256                |   3.687 |
             |____________________|_________|
             
- __learning_rate (0.0005, 0.001, 0.005)__
        - started with 0.001 then move up and down by 0.005.
        - lr = 0.001 gives slightly better result.
             ________________________________
             | Learning Rate      |    Loss |
             |____________________|_________|
             | 0.0005             |   3.677 |
             | 0.001              |   3.674 |
             | 0.005              |   4.067 |
             |____________________|_________|
        
        - alternatively, use different adaptive learning method.
 
[Source](https://towardsdatascience.com/learning-rate-schedules-and-adaptive-learning-rate-methods-for-deep-learning-2c8f433990d1): Ref: Suki Lau, Jul 29, 2017

Adaptive gradient descent algorithms such as __Adagrad, Adadelta, RMSprop, Adam__, provide an alternative to classical 
__SGD__. These per-parameter learning rate methods provide heuristic approach without requiring expensive work in tuning 
hyperparameters for the learning rate schedule manually.
<br>In brief, Adagrad performs larger updates for more sparse parameters and smaller updates for less sparse parameter. 
It has good performance with sparse data and training large-scale neural network. However, its monotonic learning rate 
usually proves too aggressive and stops learning too early when training deep neural networks. Adadelta is an extension 
of Adagrad that seeks to reduce its aggressive, monotonically decreasing learning rate. RMSprop adjusts the Adagrad 
method in a very simple way in an attempt to reduce its aggressive, monotonically decreasing learning rate. Adam is an 
update to the RMSProp optimizer which is like RMSprop with momentum.
          
- __embedding_dim (100, 150, 200)__
        - this denotes the size of embedding vector.
        - Based on the material regarding embedding, dimensions around 200 - 300 are more appropriate to be used.
        - in my case, am keeping embedding dim to 200.
             ________________________________
             | Embedding Dim      |    Loss |
             |____________________|_________|
             | 100                |   3.679 |
             | 150                |   3.655 |
             | 200                |   3.483 |
             |____________________|_________|
             
- __<font color='red'><u>hidden_dim (128, 256, 512)</u></font>__
        - Hidden dim should be reasonable as more could means overfitting. 
        - Another approach is to keep adding hidden_dim until validation loss getting worse.
        - in my case, am keeping embedding dim to 512.
             ________________________________
             | Hidden Dim         |    Loss |
             |____________________|_________|
             | 128                |   3.905 |
             | 256                |   3.773 |
             | 512                |   3.483 |
             |____________________|_________|
             
- __<font color='red'><u>n_layers (2, 3, 4)</u></font>__
        - Keeping it between 2 and 3. More layers will lead to complex model and may also tend to be overfitting.
        - Ref: Andrej Karpathy - often a 3 layer net will outperform a 2 layer net and anything deeper rarely help much.
        - am keeping n_layers = 2.
             ________________________________
             | n_layers           |    Loss |
             |____________________|_________|
             | 2                  |   3.483 |
             | 3                  |   3.866 |
             | 4                  |   5.774 |
             |____________________|_________|       

- __num of training iterations / num_epochs (10, 15)__
        - as this iteration process takes considerable amount of time, I am keeping it as last part of the tuning.
        - I started with epochs=10 and then extended to 15.
        
================================================================================================================
<br>__Final Result__:<br>
    __sequence_length = 10, batch_size = 128, learning_rate = 0.001, embedding_dim = 200, hidden_dim = 512, n_layers = 2__<br>
Training for 15 epoch(s)...<br>
Epoch:    15/15     Loss: 2.9093032450675964
<br>================================================================================================================

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [48]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [49]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [50]:
# run the cell multiple times to get different results!
gen_length = 200 # modify the length to your preference
prime_word = 'kramer' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:47: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


kramer: models.

jerry: what are you talking about?

elaine:(to jerry) i don't know.

jerry: oh, i know.

george: what?

kramer: i don't know.

jerry: well, you know, i think i should be going.

george: what?

jerry: i dont know. i know.

kramer: well, i don't know how it can be. i don't want you to get a job.

jerry: i mean i know.

elaine: well, i don't know, i'm not going to be able to do it.

elaine: oh, you know, i think it's fair. i think i may go through the doors.

elaine: what is it?

jerry: i don't know. i know. i just want to get some ointment.

kramer: yeah.

jerry: you know i was thinking maybe i should be in a while, you know, and i want you to be a good


#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [52]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.